src/items.rs

```diff
@@ -174,23 +242,80 @@ pub enum Instance {
     },
 }
 
+impl Display for Instance {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Instance::Cnf { num_vars, clauses } => {
+                Instance::fmt_cnf(f, *num_vars, clauses)
+            }
+            Instance::Sat {
+                num_vars,
+                extensions,
+                formula,
+            } => Instance::fmt_sat(f, *num_vars, extensions, formula),
+        }
+    }
+}
+
 impl Instance {
     /// Creates a new SAT instance for `.cnf` files with given clauses.
     pub fn cnf(num_vars: u64, clauses: Vec<Clause>) -> Instance {
         Instance::Cnf {
-            num_vars: num_vars,
+            num_vars,
             clauses: clauses.into_boxed_slice(),
         }
     }
 
-    /// Creates a new SAT instance for `.sat` files with given extensions and an underlying formula.
-    pub fn sat(num_vars: u64, extensions: Extensions, formula: Formula) -> Instance {
+    /// Creates a new SAT instance for `.sat` files with given extensions and
+    /// an underlying formula.
+    pub fn sat(
+        num_vars: u64,
+        extensions: Extensions,
+        formula: Formula,
+    ) -> Instance {
         Instance::Sat {
-            num_vars: num_vars,
-            extensions: extensions,
-            formula: formula,
+            num_vars,
+            extensions,
+            formula,
         }
     }
+
+    fn fmt_sat(
+        f: &mut std::fmt::Formatter<'_>,
+        num_vars: u64,
+        extensions: &Extensions,
+        formula: &Formula,
+    ) -> std::fmt::Result {
+        writeln!(f, "p {} {}", extensions, num_vars)?;
+        writeln!(f, "{}", formula)
+    }
+
+    fn fmt_cnf(
+        f: &mut std::fmt::Formatter<'_>,
+        num_vars: u64,
+        clauses: &[Clause],
+    ) -> std::fmt::Result {
+        writeln!(f, "p cnf {} {}", num_vars, clauses.len())?;
+        for clause in clauses.iter() {
+            for literal in clause.lits() {
+                write!(f, "{} ", literal)?;
+            }
+            writeln!(f, "0")?;
+        }
+        Ok(())
+    }
+
+    /// Creates a SAT or CNF instance, converting it into a String
+    /// `comments` is a list of comments which are inserted into the
+    /// beginning of the resulting String.
+    pub fn serialize(&self, comments: &[String]) -> String {
+        let comments: Vec<String> =
+            comments.iter().map(|x| format!("c {}", x)).collect();
+        let comments: String = comments.join(LINE_ENDING);
+
+        let body = self.to_string();
+        format!("{}{}{}{}", comments, LINE_ENDING, body, LINE_ENDING)
+    }
```

Comment:
This method allocates 4 times but with different API doesn't need to allocate at all.
Use
```rust
pub fn serialize<O>(&self, comments: &[str], output: &mut O) -> core::fmt::Result
where
    O: core::fmt::Write,
{
    for comment in comments {
        writeln!(output, "c {}", comment)?;
    }
    writeln!(output, {}", self)?;
}
```

Haven't checked if this compiles but at least you can use it as template.

Reply:
Made these two changes.